src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2930,7 +2930,7 @@ private sealed class VariableAssignmentVisitor : AstVisitor2
 
             private void SetLastAssignment(Ast ast, bool enumerate = false, bool redirectionAssignment = false)
             {
-                if (LastAssignmentOffset < ast.Extent.StartOffset)
+                if (LastAssignmentOffset < ast.Extent.StartOffset && !VariableTarget.Extent.IsWithin(ast.Extent))
```

Comment:
@MartinGC94 Could you please add a comment to explain why we need the `!VariableTarget.Extent.IsWithin(ast.Extent)` part to make it easier to understand? I think it'd be good to include the example `$x = Get-Random; $x = $x.Where{$_.<tab>}` in the comment. Thanks!